### ISSUE TITLE
Difference "loss": "ctc" and "ctc_loss" layer

### ISSUE THREAD

**Issue by Atticus1806** [rwth-i6/returnn#1077]
I am right now reproducing a config which is using 
```python
        'phoneme_probs': {'class': 'softmax',
                        'from': ['spectrogram_encoder'],
                        'target':['data:classes'],
                        'loss': 'ctc'}
```
as a loss layer. 
Since I am using returnn_common I defaulted back to 
```python
    spectogram_encoder = self.spectogram_lin(spectogram_encoder)
    ctc = nn.ctc_loss(logits=spectogram_encoder, targets=phonemes)
    ctc.mark_as_loss()
```
which wraps `CtcLossLayer` and expects logit input, thus no softmax here. 
The Problem now is: This returnn_common model outputs score values arround 400 in the beginning while the original model (and hand changnig the loss definition of the network) outputs values arround 2.6 .
This in general can of course converge to the same optimum and might just be scaled differently, but evaluating for TTS gives a far worse performance for the `CtcLossLayer`.   
Is there something else that needs to be changed in order to use `CtcLossLayer`? 

The rest of the network has the same amount of parameters and similar performance if I change the loss by hand.

**Comment by albertz** [comment]
`SoftmaxLayer` also has another linear transformation in it, and is not just softmax. You have that in returnn_common as well?

Another difference: `CtcLoss` by default uses the standard `tf.nn.ctc_loss`, while `CtcLossLayer` always uses `returnn.tf.native_op.ctc_loss`. You can use the option `use_native` for `CtcLoss`. In theory, it should not make a difference but our native variant should be much faster and esp not throttle the CPU. There are also a couple of test cases active to validate that. But you might want to double check. It might have very subtle effects on convergence.

In general, `CtcLossLayer` is not really well tested (or not really at all). I think we only introduced it now for returnn_common. So there might be bugs. If `CtcLoss` with `use_native` still works fine, you should more carefully check the code.

Another difference: `CtcLoss` does some clever logic to get the logits, the values before the softmax. If that fails, it calculates `safe_log(layer_output)`. In theory, it should not fail, but you might want to double check that.

**Comment by Atticus1806** [comment]
The linear transformation is done beforehand `spectogram_encoder`. 

I started a training with `use_native` and the score is equal up to 3rd decimal after first epoch (which is would neglect as noise), so that seems fine. I had to set beam_width to 1, but I guess thats fine since from documentation its only for eval. 

What about normalization / flattening? Is the code there using the same functions? CtcLossLayer is using loss as_is thus auto flattening, what about CtcLoss? For me it seems that the loss is in a different domain maybe or does not do some averaging etc.? That would explain the factor >100 difference I think.

**Comment by albertz** [comment]
> What about normalization / flattening? Is the code there using the same functions? CtcLossLayer is using loss as_is thus auto flattening, what about CtcLoss? For me it seems that the loss is in a different domain maybe or does not do some averaging etc.? That would explain the factor >100 difference I think.

Oh yes that could be the difference. In `CtcLoss`, it uses the default normalization logic from the base `Loss`, which is:
```python
class Loss:
  ...
  def init(self, output, output_with_activation=None, target=None, layer=None):
    ...
      if self.output.have_time_axis():
        ...
        if target:
          assert target.have_time_axis()
          self.target_seq_lens = target.get_sequence_lengths()
          self.target_flat = self._flatten_or_merge(target)
          self.loss_norm_factor = 1.0 / tf.cast(tf.reduce_sum(self.target_seq_lens), tf.float32)
        else:
          self.loss_norm_factor = 1.0 / tf.cast(tf.reduce_sum(self.output_seq_lens), tf.float32)
      else:  # no time axis
        ...
        if self.output.have_batch_axis():
          self.loss_norm_factor = (
            1.0 / tf.cast(tf.shape(self.output.placeholder)[self.output.batch_dim_axis], tf.float32))
        else:
          self.loss_norm_factor = 1.0
...
```

So in `CtcLoss`, it uses the target seq length for normalization. I wonder a bit if this is actually what one would expect, or if the output seq length makes more sense?

In `CtcLossLayer` with `AsIsLoss`, the output does not have a time dim. So it just normalizes over the num seqs with this code.

So, what's actually the wanted behavior?

I think `AsIsLoss` behaves as expected, it also can not really do otherwise because it does not have any other information. When it just gets a tensor of shape [B], it will just use that as loss.

`CtcLossLayer`? It just calculates the CTC loss, as it is defined originally, the textbook equation. So I would say this is also as expected.

Maybe `CtcLoss` is actually behaving strange or wrong here. But we cannot really change that now. Or only with new behavior_version.

So, I would say returnn_common is fine. Maybe use a loss scale=0.01 or so to get it similar to the `CtcLoss` behavior. I assume by tuning this it should get very similar as before. I don't expect that the `CtcLoss` behavior is really superior.

Dividing the CTC loss by the target seq length and then using that as loss is not exactly the same. This does not do proper normalization then. RETURNN does the normalization globally over the dataset, while this would only make it locally for the batch, which is not equivalent.

I think via `custom_inv_norm_factor`, which you can also use via returnn_common `mark_as_loss`, you can get the behavior. I think sth like:
```python
ctc_loss.mark_as_loss(custom_inv_norm_factor=nn.length(targets, axis=targets_time_dim))
```

**Comment by Atticus1806** [comment]
Thanks for the explanation. 
I am currently trying your `custom_inv_norm_factor` option and will report on that once I have WER results which might take a few days. 
In general if we can confirm that this is in fact the difference, then I would strongly suggest adding a note in the documentation that this is a difference between the two.
What I can already confirm is that the loss over 100 epochs of my models performs similar with both `use_native` enabled and disabled, so I think the native implementation should be fine.

**Comment by albertz** [comment]
I think this is resolved, right?

**Comment by Atticus1806** [comment]
Yes it should be. While we were not fully able to reproduce the results yet, we are confident this is not due to a difference in loss.
